docs/features/incremental-generators.md

```diff
@@ -42,154 +42,711 @@ public class MyGenerator : IIncrementalGenerator { ... }
 An assembly can contain a mix of diagnostic analyzers, source generators and
 incremental generators.
 
-### Initialization
-
-`IIncrementalGenerator` has an `Initialize` method that is called by the host
-(either the IDE or the command-line compiler) exactly once, regardless of the
-number of further compilations that may occur. `Initialize` passes an instance
-of `IncrementalGeneratorInitializationContext` which can be used by the
-generator to register a set of callbacks that affect how future generation
-passes will occur.
-
-Currently `IncrementalGeneratorInitializationContext` supports two callbacks:
+### Pipeline based execution
 
-- `RegisterForPostInitialization(GeneratorPostInitializationContext)`: the same
-  callback in shape and function as supported by Source Generators today
-- `RegisterExecutionPipeline(IncrementalGeneratorPipelineContext)`: replaces Execute, described below
+`IIncrementalGenerator` has an `Initialize` method that is called by the host[^1]
+exactly once, regardless of the number of further compilations that may occur.
 
-### Pipeline based execution
+[^1]: Such as the IDE or the command-line compiler
 
 Rather than a dedicated `Execute` method, an Incremental Generator instead
-creates an execution 'pipeline' as part of the initialization process via the
-`RegisterExecutionPipeline` method:
+defines an immutable execution pipeline as part of initialization. The
+`Initialize` method receives an instance of
+`IncrementalGeneratorInitializationContext`which is used by the generator to
+define a set of transformations:
 
 ```csharp
 public void Initialize(IncrementalGeneratorInitializationContext initContext)
 {
-    initContext.RegisterExecutionPipeline(context =>
-    {
-        // build the pipeline...
-    });
+    // define the execution pipeline here via transformations of initContext
 }
 ```
 
-This pipeline is not directly executed, but instead consists of a set of steps
-that are executed on demand as the input data to the pipeline changes. Between
-each step the data produced is cached, allowing previously calculated values to
-be reused in later computations when applicable, reducing the overall
-computation required between compilations.
+These transformations are not executed directly at initialization. Instead they
```

Comment:
Consider relating it to an existing C# feature like iterators. The iterator doesn't execute immediately, instead it defines a set of actions that occur as the caller pulls on values.